### ISSUE TITLE
qsm processing error 

### ISSUE THREAD

**Issue by abolfazl-ma** [qMRLab/qMRLab#357]
Hello qMRILab developers 
I am using your package for QSM analysis . 
when I run the code I got some errors which is not resolvable for me 
I will attach it below . 
I appreciate any help . 
thanks a lot 

Error using logical
NaN's cannot be converted to logicals.

Error in qsm_sb/fit (line 194)
  data.Mask = logical(data.Mask);

Error in FitData (line 209)
    Fit = Model.fit(data);

Error in qMRLab>FitGo_FitData (line 362)
FitResults = FitData(data,Model,1);

Error in qMRLab>FitGO_Callback (line 344)
FitGo_FitData(hObject, eventdata, handles);

Error in gui_mainfcn (line 95)
        feval(varargin{:});

Error in qMRLab (line 37)
    gui_mainfcn(gui_State, varargin{:});

Error in
matlab.graphics.internal.figfile.FigFile/read>@(hObject,eventdata)qMRLab('FitGO_Callback',hObject,eventdata,guidata(hObject)) 
Error while evaluating UIControl Callback.

**Comment by tanguyduval** [comment]
@abolfazl-ma thanks a lot for your feedback. It seems that your mask is not binary. Can you please attach your mask here?
Best,
Tanguy

**Comment by abolfazl-ma** [comment]
thanks for your response. 
yes of  course .[]
[New WinRAR ZIP ar
[New WinRAR ZIP archive.zip](https://github.com/qMRLab/qMRLab/files/3650638/New.WinRAR.ZIP.archive.zip)
chive.zip](https://github.com/qMRLab/qMRLab/files/3650635/New.WinRAR.ZIP.archive.zip)

**Comment by abolfazl-ma** [comment]
we try this with another PC and  we got the following error.

Error using iterapp (line 59)
user supplied function ==> @(x)applyForward(x,D2,lambdaL2,fdx,fdy,fdz,conj(fdx),conj(fdy),conj(fdz),magnWeight)
failed with the following error:

 To RESHAPE the number of elements must not change.

Error in pcg (line 161)
r = b - iterapp('mtimes',afun,atype,afcnstr,x,varargin{:});

Error in calcChiL2 (line 41)
    [F_chi, ~, pcg_res, pcg_iter] = pcg(@(x) applyForward(x, D2, lambdaL2, fdx, fdy, fdz, ...

Error in qsm_sb/fit (line 331)
      [FitResults.chiL2,FitResults.chiL2M] = calcChiL2(phaseLUnwrap, lambdaL2, FitOpt.direction,
      imageResolution, maskGlobal, padSize, magnWeight);

Error in FitData (line 209)
    Fit = Model.fit(data);

Error in qMRLab>FitGo_FitData (line 362)
FitResults = FitData(data,Model,1);

Error in qMRLab>FitGO_Callback (line 344)
FitGo_FitData(hObject, eventdata, handles);

Error in gui_mainfcn (line 95)
        feval(varargin{:});

Error in qMRLab (line 37)
    gui_mainfcn(gui_State, varargin{:});

Error in
matlab.graphics.internal.figfile.FigFile/read>@(hObject,eventdata)qMRLab('FitGO_Callback',hObject,eventdata,guidata(hObject)) 
Error while evaluating UIControl Callback.

best ,
abolfazl.

**Comment by tanguyduval** [comment]
@abolfazl-ma your first issue has been solved and will be merged ASAP, thanks for this feedback!
For the second issue. @mathieuboudreau will help you since he implemented this module.
Best,
Tanguy

**Comment by tanguyduval** [comment]
@mathieuboudreau it would be nice also to manage NaNs in Phase or Magnitude data. Currently, the entire result is NaN if a single input voxel is NaN.

**Comment by mathieuboudreau** [comment]
@tanguyduval thanks for your help so far!

To clear one thing first, I did not implement this code; this code is from Berkin Bilgic (it was uploaded here: https://www.martinos.org/~berkin/software.html), and the module was implemented by @agahkarakuzu. I helped @agahkarakuzu a little bit, but he did most of the work. 

To start off, I'd like to know if @abolfazl-ma you have tried to download and process the test dataset. QSM datasets come in all sorts of formats, and this module is only compatible with datasets matching the ones with the same format as that one. Because we know that the code worked at processing the test dataset, we would assume that if you've formatted your data in the same format, then it should also process. Some features of this dataset are: single echo (not multi-echo), 3D, matching dimensions for phase, magnitude, and masks volumes, masks only having values of 0 and 1, phase data should be unprocessed (wrapped) and range in values from -pi to pi, and magnitude data have only positive values.

It is very concerning that your mask has NaN (not a numbers), so I would investigate that. How did you create your masks? And does your Phase and Magnitude data also have NaN values (as Tanguy hints)? I don't think there is one clear solution in qMRLab to handling NaN (by setting them to a certain value); it should be handled at the source I think.

>@mathieuboudreau it would be nice also to manage NaNs in Phase or Magnitude data. Currently, the entire result is NaN if a single input voxel is NaN.

That make sense @tanguyduval, as unlike most of our other models, QSM relies heavily on spatial information, and neighbouring voxels depend on each other during the processing (if I recall, QSM has a deconvolution step, or an alternative implementation of that depending on the algoritm). Any NaN in the dataset is worrying, and I'm not sure how other QSM experts would handle this besides acquiring good data (as far as I know, MRI scanners shouldn't output NaN from the raw data, so NaNs would mean that maybe something went wrong in the preprocessing steps, and should be fixed there and not here).

**Comment by agahkarakuzu** [comment]
Thank you for the great points @mathieuboudreau!

I am also happy to help out @abolfazl-ma with their very first issue on GitHub (🎊) following responses to your questions.

**Comment by abolfazl-ma** [comment]
Thanks for the information you provided . 

In my data, there are no values of NAN, and in the second error we can see that this is not the case . my data is 3D multi echo . and  there are other conditions that you said . And I couldn't download the sample data .     On Wednesday, September 25, 2019, 04:28:19 PM GMT+3:30, Agah <notifications@github.com> wrote:  
 
 
Thank you for the great points @mathieuboudreau!

I am also happy to help out @abolfazl-ma with their very first issue on GitHub ( 🎊 ) following responses to your questions.

**Comment by mathieuboudreau** [comment]
> my data is 3D multi echo

Unfortunately, our sole QSM module is implemented to be only compatible with single-echo data. It is based off of [this paper](https://github.com/qMRLab/qMRLab/blob/master/src/Models/QSM/qsm_sb.m#L55-L59), and is a import/refactor of Bilgic's code that he released on his website [here](https://www.nmr.mgh.harvard.edu/~berkin/software.html) (it's the "Fast l1-Regularized QSM with..." one). This is likely what caused your error. We currently don't have any plans on implementing another QSM module, but I'll bring it up at our next team meeting and get back to you. If we do decide to implement a multi-echo compatible QSM implementation that someone else released online already, it will will likely take a while (I'd say a couple of months at least, since the ISMRM and OHBM deadlines are coming up soon).

In the meantime, may I recommend checking out this list of open-source QSM softwares & datasets that I compiled last year: https://github.com/mathieuboudreau/qsm-tools. I know that a few of these are widely used in studies already.

We'll update the documentation/data check to make it clear that our `qsm_sb` implementation requires a single echo dataset, and are compatible with multi-echo datasets.

p.s. 

>And I couldn't download the sample data .

That's odd. Did you get an error? If you clicked the button and selected a folder, it should have downloaded it, it may just not appear until you click "View" next to Phase or Magnitude paths.

**Comment by agahkarakuzu** [comment]
@mathieuboudreau I remember we were discussing with a contributor during MRathon about a pre-processing step to combine multi echo data before proceeding to perform qsm. I'll check that and see if what we were talking there had anything to do with this.

**Comment by jeremie-fouquet** [comment]
Hi all! I'm doing fine, thanks. Hope all of you too!

Ok, so if I understand correctly, I indeed think that part of the actual issue is related to what we discussed at the last MRathon. We could indeed make qMRLab QSM module compatible with multi-echo data sets by adding a preprocessing module where multi-echo data would be combined into an SNR-optimized estimate of the frequency. This frequency estimate could then be input to the rest of the QSM module as if it had been estimated from a single-echo data set. Based on my current knowledge, this is the most common way to process multi-echo data sets in the QSM community at the moment (although other methods exist and are being developed and tested).

At the last MRAthon, I wanted to implement one of the two main methods used to combine multi-echo data sets (by the way, sorry not to have given any news sooner...). It's the weighted average method, explained in [this paper](doi.org/10.1016/j.neuroimage.2011.07.019). The qMRLab implementation could be ready rather rapidly, I was not that far from a working version. 

However, I paused the development because talking and coding with you guys made me realize a quite important thing. I don't know if you remember, but we decided to place the weighted average echo combination module into the B0 estimation module because it essentially aims at computing a frequency estimate from phase data. However, I realized that the weighted-average approach does not exactly yield the frequency: the result is still contaminated with the phase offset (although this was not discussed in the original paper).  While this can be harmless in a QSM setup where low spatial frequencies are mostly removed during the background field removal step, it causes an important problem for the B0 estimation module. Therefore, I was a bit stuck and did not know exactly how to implement the echo combination module...

Now that we are at it, I can propose the two more feasible solutions that I thought of:
1. Keep the weighted average in the B0 estimation module, but estimate the phase offset alongside the frequency (I only have simple non-tested methods in mind for that at the moment).
2. Code the weighted average exactly as it was proposed in the original paper, but do not include it in the B0 estimation module.

Other options in between those two are also feasible... 

So, in short, I would be happy to continue and finish the implementation of the weighted average multi-echo preprocessing module. To do so, it would be helpful to have your thoughts about the previous alternatives!

**Comment by agahkarakuzu** [comment]
Thank you for the quick response @jeremie-fouquet! What I have mind for this one is actually to have a separate module that inherits the B0 module. I think model names should reveal their functionality clearly for users. 

> Code the weighted average exactly as it was proposed in the original paper, but do not include it in the B0 estimation module

If it is not possible to implement it  as you mention here by inheriting from B0, then we can indeed add it as a new model.

**Comment by agahkarakuzu** [comment]
@jeremie-fouquet I've just merged `master` into `echo_combination`, was this the branch you were working on? How quickly can we come up with something that functions to combine multiple echoes? 

If implementing is qMRLab is gonna take you a while I can also help and/or/if possible can you share your echo combination code so that we can help out @abolfazl-ma more quickly?

**Comment by jeremie-fouquet** [comment]
That was indeed the branch I was working on. We can come up with something that combines echoes rather quickly (a few days), the algorithm is quite simple. I could push something functioning something like tomorrow, but I will probably need some help then to make it fit properly within qMRLab.

**Comment by agahkarakuzu** [comment]
@jeremie-fouquet thank you 🚀 Don't worry about the qMRLab related aspects of it, I can arrange them quickly. As long as you put your code out there with clear input/output/parameter selection explanations, the rest is going to be easy.

**Comment by jeremie-fouquet** [comment]
Hi @agahkarakuzu . I've pushed a first version of the weighted average echo combination on my fork [here](https://github.com/jeremie-fouquet/qMRLab/tree/echo_combination). qsm_sb should now support multi-echo data sets, with weighted averaging when magnitude is available and simple averaging when it is not.

I would need help to see what's the best way to allow someone to enter its own TE vector (length a priori unknown) in the GUI.

The function performing the weighted average still needs testing and improvements. I'm working on these improvements at the moment. @abolfazl-ma, in the meantime, if you prefer an echo combination matlab implementation that has been more widely used in the community, you can give a try to the Fit_ppm_complex_TE.m in the [MEDI toolbox](http://pre.weill.cornell.edu/mri/pages/qsm.html).

I also allowed myself to do a little bit of refactoring to qsm_sb.

**Comment by agahkarakuzu** [comment]
Thank you @jeremie-fouquet, I'll look into this today. 

> I also allowed myself to do a little bit of refactoring to qsm_sb.

Why do you think that a refactoring is needed?

**Comment by jeremie-fouquet** [comment]
> Why do you think that a refactoring is needed?

Mainly, the scaling of the unwrapped phase to ppm units was done in the SHARP function. This caused a small issue because part of this scaling is now done in the weighted average part (namely the scaling with 1/TE). 

Because I don't think scaling to ppm units is inherent to SHARP background field removal, I just moved it to the main qsm_sb body.

**Comment by mathieuboudreau** [comment]
In particular, why remove the physical parameters from the function `backgroundRemovalSharp`? (These were included by Berkin Bilgic in his original implementation). The only worry I might have is that his original results might not replicate in qMRLab now (which they did previously)

**Comment by mathieuboudreau** [comment]
But I'm not fully attached to the either way, reproducibility of his results was just a good benefit; it's something that should be tested.

**Comment by jeremie-fouquet** [comment]
>But I'm not fully attached to the either way, reproducibility of his results was just a good benefit; it's something that should be tested.

I fully agree that reproducibility is key for a simple refactoring. If I knew how I would have tested reproducibility before pushing changes. Is there a simple way to do this locally? I could run qsm_sb on the test data from let's say the master branch, then run it from the new echo_combination branch, and compare the results obtained in both situations, but this appears a little cumbersome. Is there an easier way to do this test within qMRLab?

**Comment by jeremie-fouquet** [comment]
> In particular, why remove the physical parameters from the function `backgroundRemovalSharp`?

The practical reason is that part of what was done with those physical parameters within `backgroundRemovalSharp `is now done within the echo combination step. Particularly, the phase was scaled with 1/TE in `backgroundRemovalSharp`. If you have multiple TE, then this does not really make sense. The idea is now to take care of this scaling with 1/TE in the echo combination part. 

A more fundamental reason is that the physical parameters are not needed for SHARP background field removal. I would guess that Berkin Bilgic included scaling with those parameters in there because it's not a complicated step that can be included easily pretty much everywhere...

**Comment by agahkarakuzu** [comment]
@jeremie-fouquet I run `qsm_sb` from your branch vs `master` on test data, then compared outputs:

![image](https://user-images.githubusercontent.com/9632322/66224406-c626cc80-e6a3-11e9-93f4-a285335fba88.png)

**Comment by jeremie-fouquet** [comment]
I'll have a look. Could you please indicate how you obtain (new/old) nfm and sb?

**Comment by agahkarakuzu** [comment]
1. Download test data from [here](https://osf.io/549ke/download?version=4)
2. For new, switch to the `echo_combination` branch.
3. For old, switch to the `master` branch. 

`nfm` indicates that `No Regularization` option was selected. 
`sb` indicates that `Split-bregman` option was selected. 

Other (fixed) options: 

`Derivative direction`: Forward
`Sharp Filtering` : Enabled 
`Sharp Mode`: once 
`Padding size`: 9 9 9
`LambdaL1`: 0.000921055 (default)
`LambdaL2`: 0.0316228 (default)

**Comment by jeremie-fouquet** [comment]
Ok. So this seems to come from the fact that, before refactoring, data was padded for SHARP with `padVolumeForSharp` prior to phase unwrapping. Honestly, I don't know if that was right or wrong. I just know that padding is not needed for phase unwrapping.

Because the echo combination returns an already unwrapped phase, I moved data padding after phase unwrapping during the refactoring (it is unnecessary to pad data for echo combination and phase unwrapping).

However, it turns out that padding before or after unwrapping does not make that much of a difference. To illustrate my point, here is the difference in nfm chi between master and echo_combination branch as they are now in the qMRLab/qMRLab (padding is done after unwrapping):
![chi_nfm_pad_after_uw](https://user-images.githubusercontent.com/6575734/66233568-b0240680-e6b9-11e9-8f77-0021fd53d5d1.png)

Here is the difference between the same branches, but after I've made modifications to echo_combine so that padding occurs before phase unwrapping (again, I don't think this is necessary; actually it might even make phase unwrapping sensitive to the choice of padding made for SHARP, which I don't think is desirable):
![chi_nfm_pad_before_uw](https://user-images.githubusercontent.com/6575734/66233663-dea1e180-e6b9-11e9-8c8e-2c2405ca10e6.png)


Please note the scale of the differences. In the first case, assertElementsAlmostEqual returns an error. It does not in the second.

**Comment by agahkarakuzu** [comment]
@jeremie-fouquet thank you for the clear explanation! I'll merge your `echo_combination` again to have these changes. e-16 is quite acceptable :)

**Comment by jeremie-fouquet** [comment]
Indeed, haven't pushed them yet! I thought you might have wanted to consider the option of padding after unwrapping instead of before (even if it brings ~e-6 difference with previous version). If you want to minimize difference with before, I'll push the changes as soon as I can (maybe tonight).

**Comment by jeremie-fouquet** [comment]
But honestly, I think I would consider unwrapping before padding (i.e., what is done in the actual version of the qMRLab/echo_combination branch), unless there is a good reason not to do so that I don't get yet. This would make unwrapping independent of the padding chosen for SHARP.

**Comment by agahkarakuzu** [comment]
@jeremie-fouquet have you pushed changes you mentioned above? Unwrapping --> padding order intuitively makes sense, I totally get your point. But just for the sake of reproducing the original implementation I'd say we leave as is. The one that gives this output:

![image](https://user-images.githubusercontent.com/9632322/66366243-c6ea8780-e95d-11e9-8bdd-78f9fd29d513.png)

Many thanks 🌟 🚀 🎊

**Comment by jeremie-fouquet** [comment]
@agahkarakuzu All right, it is now pushed on my echo_combination branch. I hope everything works fine.

**Comment by agahkarakuzu** [comment]
Many thanks @jeremie-fouquet, merging rn ! I think now @abolfazl-ma can give it a try on the `echo_combination` branch right?

**Comment by jeremie-fouquet** [comment]
Actually, I think one thing lacking is the possibility for the user to enter its own TE vector (instead of a single TE). Would that be easy to do within the actual qsm_sb GUI?

**Comment by agahkarakuzu** [comment]
Great reminder @jeremie-fouquet , I will do it :) 

- [ ] Add multi-echo option on GUI

**Comment by agahkarakuzu** [comment]
@jeremie-fouquet if you have time, can you try using add button below to pass multiple echo times and use the multiecho data you've been working with? 

![image](https://user-images.githubusercontent.com/9632322/66406333-743eb900-e9b9-11e9-9e6d-f0a10449f9a8.png)

**Comment by jeremie-fouquet** [comment]
@agahkarakuzu Seems to work for me. I also had to modify my echo_combination branch for a few remaining bugs (the new modifs are pushed).

**Comment by agahkarakuzu** [comment]
@abolfazl-ma thanks to great support of @jeremie-fouquet now our `qsm_sb` module can work with multi-echo data.  We'd be glad if you can give it a try and let us know if it works for you? 

![image](https://user-images.githubusercontent.com/9632322/66432038-2f337a80-e9eb-11e9-9113-a821daaf4474.png)

Please add a `TE (s)` row for every echo and enter the corresponding echo time. You can use `Add` button to do that.

**Comment by agahkarakuzu** [comment]
@abolfazl-ma are you planning to give the module a try and see if it works for you? We'd like to know whether the implementations made responded your need.

**Comment by agahkarakuzu** [comment]
@abolfazl-ma the changes @jeremie-fouquet did on this branch appear to resolve the issue you were facing. 

Please note that we will merge changes to the master soon. It would be really useful to have your feedback as well if you are still on it.

**Comment by agahkarakuzu** [comment]
#346
